packages/hardhat-core/scripts/console-library-generator.ts

```diff
@@ -1,44 +1,131 @@
+  // We can think of a permutation as a number of base `elemCount`, i.e.
+  // each 'digit' is a number between 0 and `elemCount - 1`.
+  // Then, to generate all permutations, we simply need to linearly iterate
+  // from 0 to max number of permutations (elemCount ** len) and convert
+  // each number to a list of digits as per the base `elemCount`, see above.
+  const numberOfPermutations = elemCount ** len;
+  const dividers = Array(elemCount)
+    .fill(0)
+    .map((_, i) => elemCount ** i);
+
+  for (let number = 0; number < numberOfPermutations; number++) {
+    const params = Array(len)
+      .fill(0)
+      .map((_, i) => Math.floor(number / dividers[i]) % elemCount);
+    // Reverse, so that we keep the natural big-endian ordering, i.e.
+    // [0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], ...
+    params.reverse();
+
+    yield params;
+  }
+}
+
+type TypeName = { type: string; modifier?: "memory" };
+type FnParam = TypeName & { name: string };
+
+/** Computes the function selector for the given function with simple arguments. */
+function selector({ name = "", params = [] as TypeName[] }) {
+  const sigParams = params.map((p) => p.type).join(",");
+  return keccak256(Buffer.from(`${name}(${sigParams})`)).slice(0, 4);
+}
+
+function toHex(value: Uint8Array) {
+  return "0x" + Buffer.from(value).toString("hex");
 }
 
-const types = ["uint256", "string memory", "bool", "address"];
+/** The types for which we generate `logUint`, `logString`, etc. */
+const SINGLE_TYPES = [
+  { type: "int256" },
+  { type: "uint256" },
+  { type: "string", modifier: "memory" },
+  { type: "bool" },
+  { type: "address" },
+  { type: "bytes", modifier: "memory" },
+  ...Array.from({ length: 32 }, (_, i) => ({ type: `bytes${i + 1}` })),
+] as const;
+
+/** The types for which we generate a `log` function with all possible
+ combinations of up to 4 arguments. */
+const TYPES = [
+  { type: "uint256" },
+  { type: "string", modifier: "memory" },
+  { type: "bool" },
+  { type: "address" },
+] as const;
+
+/** A list of `console.log*` functions that we want to generate. */
+const CONSOLE_LOG_FUNCTIONS =
+  // Basic `log()` function
+  [{ name: "log", params: [] as FnParam[] }]
+    // Generate single parameter functions that are type-suffixed for
+    // backwards-compatibility, e.g. logInt, logUint, logString, etc.
+    .concat(
+      SINGLE_TYPES.map((single) => {
+        const param = { ...single, name: "p0" };
+        const nameSuffix = capitalize(param.type.replace("int256", "int"));
+
+        return {
+          name: `log${nameSuffix}`,
+          params: [param],
+        };
+      })
+    )
+    // Also generate the function definitions for `log` for permutations of
+    // up to 4 parameters using the `types` (uint256, string, bool, address).
+    .concat(
+      [...Array(4)].flatMap((_, paramCount) => {
+        return Array.from(
+          genPermutations(TYPES.length, paramCount + 1),
```

Comment:
```suggestion
      [...Array(5)].flatMap((_, paramCount) => {
        return Array.from(
          genPermutations(TYPES.length, paramCount),
```

As a side-note, this base case can handle generating the empty `log()` just fine but shifted its order in the console.sol file. This is probably meaningless but I wanted to avoid making any changes; let me know if you'd like me to not explicitly handle `log()` and handle this with this code, instead.

Reply:
That's a good point, but yes, let's leave it as is.